### PR TITLE
fix(tui): guard AskUserQuestion approval against malformed questions/options

### PR DESCRIPTION
**Pull request by mruderman**
## Problem

A malformed `AskUserQuestion` tool call can brick the TUI on startup with:

```
ERROR  Spread syntax requires ...iterable not be null or undefined
…/letta.js:455593
  …currentQuestion.options,
```

`ApprovalSwitch.getQuestions()` parsed args with a bare type assertion and **no runtime validation**:

```ts
function getQuestions(approval: ApprovalRequest): Question[] {
  try {
    const args = JSON.parse(approval.toolArgs || "{}");
    return (args.questions as Question[]) || [];   // ← no shape check
  } catch { return []; }
}
```

When a model emits `questions` as a **JSON string** instead of an array (observed in the wild — a gstack `/office-hours`-style skill produced exactly this), that string is returned as-is. Then:

- `ApprovalSwitch` checks `questions.length > 0` → a string has `.length`, so it **passes** → renders `InlineQuestionApproval`.
- `InlineQuestionApproval` does `currentQuestion = questions[0]` → the first **character** (truthy) → `...currentQuestion.options` → `undefined` → **throws**.
- The malformed tool call is persisted as a pending approval, so the TUI **crashes on every subsequent startup** until the approval is cleared (`letta --new`) or denied.

`options` is required by the tool schema, but the renderer must not assume well-formed input — it runs before the tool implementation can reject bad args.

## Fix (two layers)

1. **`getQuestions()` validates shape** — returns `[]` unless `questions` is a non-empty array where every entry is an object with a non-empty `options` array. Malformed payloads fall through to `InlineGenericApproval`, matching how malformed Bash/Task args are already handled.
2. **`InlineQuestionApproval` guards `currentQuestion.options`** with `?? []` as defense-in-depth for any other path that could construct the component with bad data.

## Verification

- New `ApprovalSwitch.test.ts` (9 cases): string-encoded `questions`, missing/null/empty `options`, non-array, partial malformation, unparseable/empty args — all return `[]`; well-formed payload returns the questions.
- `bun test` ✅ 9 pass
- `tsc --noEmit` ✅ clean
- `biome check` ✅ clean

## Notes

- `getQuestions` is now `export`ed so the parser is unit-testable in isolation.
- Minimal diff: ~25 lines across two source files + a colocated test.

### CODE REVIEW

File: src/cli/components/InlineQuestionApproval.tsx [hunk 1]
```diff
@@ -72,7 +72,7 @@ export const InlineQuestionApproval = memo(
     const showOther = currentQuestion?.allowOther !== false;
     const baseOptions = currentQuestion
       ? [
-          ...currentQuestion.options,
+          ...(currentQuestion.options ?? []),
           ...(showOther ? [{ label: "Type something.", description: "" }] : []),
         ]
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/components/InlineQuestionApproval.tsx:86]: Good catch — agreed. `?? []` only covers null/undefined, so a non-iterable `options` (e.g. `{}`) would still throw on spread. Switched to `Array.isArray(currentQuestion.options) ? currentQuestion.options : []` in 46225e62, and added a render-level regression test in `InlineQuestionApproval.test.tsx` that renders with `undefined | null | {} | 42 | "nope"` options and asserts no throw.

File: src/cli/components/ApprovalSwitch.tsx [hunk 2]
```diff
@@ -325,11 +325,26 @@ function getMemoryFileEditInfo(
   }
 }
 
-// Parse questions from AskUserQuestion args
-function getQuestions(approval: ApprovalRequest): Question[] {
+// Parse questions from AskUserQuestion args.
+// Validates shape defensively: a malformed tool call (e.g. `questions` emitted
+// as a JSON string rather than an array, or a question missing its required
+// `options` array) must never reach InlineQuestionApproval, which spreads
+// `options` without a guard and would crash the TUI. Returns [] for anything
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/components/ApprovalSwitch.tsx]: Right, the wording was stale. Adopted your suggestion and refined it in 46225e62 to describe the layered defense accurately: `getQuestions` rejects malformed payloads at the boundary, and `InlineQuestionApproval` coerces `options` to an array as defense-in-depth (so it no longer throws on non-iterables either).

File: src/cli/components/approval-switch.test.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
+import { describe, expect, test } from "bun:test";
+import type { ApprovalRequest } from "@/cli/helpers/stream";
+import { getQuestions } from "./ApprovalSwitch";
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/components/approval-switch.test.ts:3]: Correct — `.ts` files must be kebab-case and this would have failed `check-filename-casing`. Renamed to `approval-switch.test.ts` in 46225e62; the casing check now passes. (The `InlineQuestionApproval.test.tsx` render test stays PascalCase since `.tsx` is exempt.)

File: src/cli/components/ApprovalSwitch.tsx [hunk 4]
```diff
@@ -325,11 +325,26 @@ function getMemoryFileEditInfo(
...
-    return (args.questions as Question[]) || [];
+    const questions = args.questions;
+    if (!Array.isArray(questions) || questions.length === 0) return [];
+    const wellFormed = questions.every(
+      (q) =>
+        q != null &&
+        typeof q === "object" &&
+        Array.isArray((q as Question).options) &&
+        (q as Question).options.length > 0,
+    );
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/components/ApprovalSwitch.tsx]: Good point — adopted. `getQuestions` now validates each entry too: a question is well-formed only if every option is a non-null object with a non-empty string `label` (so `options: [null]`, `[42]`, `[{label:""}]`, `[{description:"x"}]` all fall through to InlineGenericApproval). Added boundary tests in `approval-switch.test.ts` covering null/non-object/empty-label/missing-label/mixed entries. Commit f477f15c.

File: src/cli/components/InlineQuestionApproval.tsx [hunk 5]
```diff
@@ -72,7 +72,9 @@ export const InlineQuestionApproval = memo(
     const showOther = currentQuestion?.allowOther !== false;
     const baseOptions = currentQuestion
       ? [
-          ...currentQuestion.options,
+          ...(Array.isArray(currentQuestion.options)
+            ? currentQuestion.options
+            : []),
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/components/InlineQuestionApproval.tsx:84]: Agreed — the array-type guard didn't protect against bad entries. Added a `.filter((o): o is QuestionOption => o != null && typeof o === "object")` on the spread so non-object entries (e.g. `options: [null]`, `[42]`, or a mixed array) can't reach the render path that derefs `option.label`/`option.description`. Added render-level cases in `InlineQuestionApproval.test.tsx` for `[null]`, `[42]`, and mixed arrays → asserts no throw. Commit f477f15c.

File: src/cli/components/ApprovalSwitch.tsx [hunk 6]
```diff
@@ -325,11 +325,31 @@ function getMemoryFileEditInfo(
...
+      if (!Array.isArray(options) || options.length === 0) return false;
+      return options.every(
+        (o) =>
+          o != null &&
+          typeof o === "object" &&
+          typeof (o as { label?: unknown }).label === "string" &&
+          (o as { label: string }).label.length > 0,
+      );
+    });
+    return wellFormed ? (questions as Question[]) : [];
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/components/ApprovalSwitch.tsx]: Good catch — adopted, and extended to all rendered fields. Extracted a single validated parser `parseAskUserQuestions` (src/cli/helpers/ask-user-questions.ts) that checks: `question`/`header` are strings, `multiSelect` is boolean, `allowOther` is boolean-or-omitted, and each option has non-empty string `label` + string `description`. `getQuestions` now delegates to it. I also found a parallel unguarded parser (`getQuestionsFromApproval` in src/cli/app/approval-questions.ts, same bare `as Array<{...}>` assertion) and routed it through the same validator — one source of truth, no dual-validation drift. Commit f12a78fb.

File: src/cli/components/InlineQuestionApproval.tsx [hunk 7]
```diff
@@ -72,7 +72,11 @@ export const InlineQuestionApproval = memo(
     const showOther = currentQuestion?.allowOther !== false;
     const baseOptions = currentQuestion
       ? [
-          ...currentQuestion.options,
+          ...(Array.isArray(currentQuestion.options)
+            ? currentQuestion.options.filter(
+                (o): o is QuestionOption => o != null && typeof o === "object",
+              )
+            : []),
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/components/InlineQuestionApproval.tsx:84]: Agreed — adopted your suggested filter. The renderer now filters to entries where both `label` and `description` are strings (`typeof (o as {label?:unknown}).label === "string" && typeof (o as {description?:unknown}).description === "string"`), so `{label:{},description:""}` / `{label:"A",description:{}}` can't reach the React render path. Added render-level cases for both in InlineQuestionApproval.test.tsx. Commit f12a78fb.

File: src/cli/components/approval-switch.test.ts [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+          question: "Q?",
+          options: [{ description: "x" }] as unknown as {
+            label: string;
+            description: string;
+          }[],
+        },
+      ],
+    });
+    expect(getQuestions(approval(missingLabel))).toEqual([]);
+    // One bad entry invalidates the whole question (all-or-nothing).
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/components/approval-switch.test.ts:162]: Added — the `description: {}` case is now in approval-switch.test.ts (returns []). I also added boundary cases for non-string `question`/`header`, non-boolean `multiSelect`/`allowOther`, and missing `multiSelect`, plus a colocated approval-questions.test.ts covering the second parser (now sharing the same validator). Commit f12a78fb.

File: src/cli/helpers/ask-user-questions.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+ */
+export function parseAskUserQuestions(
+  approval: ApprovalRequest,
+): AskUserQuestion[] {
+  const parsed = safeJsonParseOr<Record<string, unknown>>(
+    approval.toolArgs,
+    {},
+  );
+  const questions = parsed.questions;
+  if (!Array.isArray(questions) || questions.length === 0) return [];
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/helpers/ask-user-questions.ts]: Good catch — real crash path. `toolArgs === "null"` parses to `null` via `safeJsonParseOr`, and `parsed.questions` threw `Cannot read properties of null`. The auto-fix you committed (da984316) guards `parsed == null || typeof !== "object" || Array.isArray(parsed)` — cleaner than my draft (uses `<unknown>` instead of a `<Record>` cast, and rejects arrays explicitly). I added the missing regression test coverage on top (4b6e45bd): null/true/42/"a string" toolArgs → [] in both approval-switch.test.ts and approval-questions.test.ts.

File: src/cli/components/approval-switch.test.ts [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+    expect(getQuestions(approval(mixed))).toEqual([]);
+  });
+
+  test("returns [] for unparseable toolArgs", () => {
+    expect(getQuestions(approval("{not valid json"))).toEqual([]);
+  });
+
+  test("returns [] for empty toolArgs", () => {
+    expect(getQuestions(approval(""))).toEqual([]);
+  });
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/components/approval-switch.test.ts:354]: Added — see 4b6e45bd. Covers toolArgs that parse to null (and true/42/string) in both the ApprovalSwitch and approval-questions parser paths (they share parseAskUserQuestions).

File: src/cli/helpers/ask-user-questions.ts [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+  // before dereferencing `.questions` — this parser runs in the render path and
+  // a throw bricks the TUI. Reject null, primitives, and arrays.
+  if (parsed == null || typeof parsed !== "object" || Array.isArray(parsed)) {
+    return [];
+  }
+  const questions = (parsed as Record<string, unknown>).questions;
+  if (!Array.isArray(questions) || questions.length === 0) return [];
+  return questions.every(isWellFormedQuestion)
+    ? (questions as AskUserQuestion[])
+    : [];
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/helpers/ask-user-questions.ts:55]: Adopted — and extended to the full contract. Verified against `src/tools/schemas/AskUserQuestion.json` (`questions.minItems=1, maxItems=4`) and `src/tools/impl/ask-user-question.ts` (throws on `length > 4`). parseAskUserQuestions now rejects `questions.length < 1 || > 4` with a comment citing the schema as the source of truth. Commit 37be3637.

File: src/cli/helpers/ask-user-questions.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  if (typeof question.multiSelect !== "boolean") return false;
+  if (
+    question.allowOther !== undefined &&
+    typeof question.allowOther !== "boolean"
+  ) {
+    return false;
+  }
+  const options = question.options;
+  if (!Array.isArray(options) || options.length === 0) return false;
+  return options.every(isWellFormedOption);
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/helpers/ask-user-questions.ts:75]: Adopted — same point, options side. Verified the schema (`options.minItems=2, maxItems=4`) and tool impl (throws on `< 2 || > 4`). isWellFormedQuestion now enforces `options.length < 2 || > 4 → reject`. This also closes the UX gap you described: we no longer render a specialized prompt the user can't successfully answer. Commit 37be3637.

File: src/cli/app/approval-questions.test.ts [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  test("returns parsed questions for a well-formed payload", () => {
+    const args = JSON.stringify({
+      questions: [
+        {
+          header: "H",
+          question: "Q?",
+          options: [{ label: "A", description: "d" }],
+          multiSelect: false,
+        },
+      ],
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/app/approval-questions.test.ts:28]: Fixed — the well-formed fixture now uses 2 options to stay in-contract with the tool schema. Commit 37be3637.

File: src/cli/components/approval-switch.test.ts [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+    // allowOther omitted is fine (it's optional).
+    const validWithoutAllowOther = JSON.stringify({
+      questions: [
+        {
+          header: "H",
+          question: "Q?",
+          options: [{ label: "A", description: "d" }],
+          multiSelect: false,
+        },
+      ],
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/components/approval-switch.test.ts:310]: Fixed — the `validWithoutAllowOther` fixture now uses 2 options. I also added a dedicated bounds regression test (>4 questions, 4 OK, 1 option, 5 options) so the contract can't silently drift. Commit 37be3637.

File: src/cli/components/approval-switch.test.ts [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+    });
+    expect(getQuestions(approval(fiveOptions))).toEqual([]);
+  });
+
+  test("returns [] when question-level fields have the wrong type", () => {
+    // InlineQuestionApproval calls `question.includes(...)` and renders
+    // `header` as a React child, so non-string values throw. multiSelect/
+    // allowOther must be booleans (allowOther may be omitted).
+    const baseOptions = [{ label: "A", description: "" }];
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/components/approval-switch.test.ts]: Fixed in 420670c7 — `baseOptions` for the question-level type tests now has 2 options, so a failure is attributable to the wrong-type field (question/header/multiSelect/allowOther), not the 2–4 count bound.

File: src/cli/components/approval-switch.test.ts [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+  test("returns [] if any single question is malformed (all-or-nothing)", () => {
+    const mixed = JSON.stringify({
+      questions: [
+        {
+          header: "H",
+          question: "Q?",
+          options: [{ label: "A", description: "" }],
+        },
+        { header: "H2", question: "Q2?" }, // missing options
+      ],
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/components/approval-switch.test.ts:332]: Good point — fixed in 420670c7. The all-or-nothing fixture's first question is now fully well-formed (2 options + multiSelect); only the second is malformed (missing options). So the test now actually pins "reject the whole batch because of one bad entry" rather than passing because every question was independently invalid.

File: src/cli/app/approval-questions.test.ts [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+          options: [{ label: "A", description: "" }],
+          multiSelect: false,
+        },
+      ]),
+    });
+    expect(getQuestionsFromApproval(approval(malformed))).toEqual([]);
+  });
+
+  test("returns [] for non-string question/header/description", () => {
+    const baseOptions = [{ label: "A", description: "" }];
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/app/approval-questions.test.ts]: Fixed in 420670c7 — `baseOptions` in approval-questions.test.ts is now 2 options, so the non-string question/header tests fail on the type check, not the count bound.

File: src/cli/app/approval-questions.test.ts [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        {
+          header: "H",
+          question: "Q?",
+          options: [
+            { label: "A", description: {} } as unknown as {
+              label: string;
+              description: string;
+            },
+          ],
+          multiSelect: false,
```
[the remark this answers is not in the document]
Comment (mruderman) [reply src/cli/app/approval-questions.test.ts:77]: Fixed in 420670c7 — the `nonStringDescription` fixture now has 2 options (one valid, one with the non-string `description`), so the failure is attributable to the `description` type, not the 2–4 bound.
